Android/BluetoothChat/app/src/main/java/ditto/live/bluetoothchat/services/BluetoothChatService.kt

```diff
@@ -0,0 +1,148 @@
+
+
+import android.app.Application
+import android.bluetooth.*
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import ditto.live.bluetoothchat.*
+
+
+object BluetoothChatService {
+
+    private val _messages = MutableLiveData<BLEMessage>()
+    val messages = _messages as LiveData<BLEMessage>
+
+    private var gattServer: BluetoothGattServer? = null
+    private var gattServerCallback: BluetoothGattServerCallback? = null
+    private var gattClient: BluetoothGatt? = null
+    private var gattClientCallback: BluetoothGattCallback? = null
+    private var gatt: BluetoothGatt? = null
+    private var gattMessageCharacteristic: BluetoothGattCharacteristic? = null
+
+    var currentDevice: BluetoothDevice? = null
+
+    /**
+     * A device has been selected in our JoinFragment
+     * Set our current device and connect our devive with the current device
+     */
+    fun setCurrentDevice(context: Application, device: BluetoothDevice) {
+        currentDevice = device
+        connectDevices(context, device)
+    }
+
+    /**
+     * Create The GattClientCallback to handle Bluetooth GATT callbacks
+     * Set our GATT client to the remote device
+     */
+    private fun connectDevices(context: Application, device: BluetoothDevice) {
+        gattClientCallback = GattClientCallback()
+        gattClient = device.connectGatt(context, false, gattClientCallback)
+    }
+
+    /**
+     * Function to write messages to the remote device
+     */
+    fun sendMessage(message: String): Boolean {
+        gattMessageCharacteristic?.let { characteristic ->
+            characteristic.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
+            val messageBytes = message.toByteArray(Charsets.UTF_8)
+            characteristic.value = messageBytes
+            gatt?.let {
+                val success = it.writeCharacteristic(gattMessageCharacteristic)
+                if (success) {
+                    _messages.value = LocalMessage(message)
+                }
+            }
+        }
+        return false
+    }
+
+    /**
+     * Function to open a GATT server
+     */
+    fun createServer(bluetoothManager: BluetoothManager, app: Application) {
+        gattServerCallback = GattServerCallback()
+        gattServer = bluetoothManager.openGattServer(
+            app,
+            gattServerCallback
+        ).apply {
+            addService(setupGattService())
+        }
+    }
+
+    /**
+     * Function to create the GATT service
+     * Set our custom service and characteristics
+     */
+    private fun setupGattService(): BluetoothGattService {
+        val gattService = BluetoothGattService(chatDiscoveryServiceID, BluetoothGattService.SERVICE_TYPE_PRIMARY)
+        val gattCharacteristic = BluetoothGattCharacteristic(
+            chatMessageCharacteristicID,
+            BluetoothGattCharacteristic.PROPERTY_WRITE,
+            BluetoothGattCharacteristic.PERMISSION_WRITE
+        )
```

Comment:
Is the app intended to be compatible with the iOS one? The GATT hierarchy is different so I don't think it would work.

Specifically, iOS uses one service UUID to locate other peers (discovery phase), then a second UUID to negotiate a chat connection. Then it has a single characteristic to exchange/notify chat message, both for write & notify.